### PR TITLE
Add quick remote view unit test

### PR DESCRIPTION
**Pull request by pasnox**
Also added base classes for unit testing.

### CODE REVIEW

File: tests/actiontest.cpp [hunk 1]
```diff
@@ -169,9 +152,9 @@ private slots:
         QCOMPARE(a1->isChecked(), false);
         QCOMPARE(changeSpy.size(), 2);
     }
-
 };
 
 QTEST_MAIN(ActionTest)
 
+#include "moc_baseprobetest.cpp"
```
Comment (vkrause) [inline tests/actiontest.cpp]: I very much like the idea. Can we compile this only once like we do for the probe object, and add this via CMake though, especially since there is now the gammaray_add_probe_test macro? Would avoid the extra moc include too.

File: core/remoteviewserver.cpp [hunk 2]
```diff
@@ -228,5 +228,6 @@ void RemoteViewServer::clientConnectedChanged(bool connected)
 void RemoteViewServer::requestUpdateTimeout()
 {
     emit requestUpdate();
+    emit frameRequested();
```
Comment (vkrause) [inline core/remoteviewserver.cpp]: Why the extra signal, isn't requestUpdate() emitted at exactly the same time? Testing that would even be better, as that's what will be what the using code will follow.

File: tests/quickinspectortest.cpp [hunk 3]
```diff
@@ -225,6 +226,93 @@ private slots:
...
+        QVERIFY(rootItem);
+        Probe::instance()->selectObject(rootItem, QPoint());
+
+        connect(remoteView, SIGNAL(frameUpdated(GammaRay::RemoteViewFrame)), remoteView, SLOT(clientViewUpdated()));
+
+        // RemoteViewServer throttle the requests with an interval of ~16ms by default
+        qreal throttlerInterval = 100.0;
+        // Our animation properties
+        qreal animationInterval = 0.0;
+        qreal animationDuration = 1400.0;
```
Comment (vkrause) [inline tests/quickinspectortest.cpp]: Testing for real frame timings looks extremely brittle, wouldn't this randomly break on high CI load?

File: core/remoteviewserver.cpp [hunk 4]
```diff
@@ -54,7 +54,9 @@ RemoteViewServer::RemoteViewServer(const QString &name, QObject *parent)
                                                     name), this, "clientConnectedChanged");
 
     m_updateTimer->setSingleShot(true);
-    m_updateTimer->setInterval(10);
+    // Qml render @ ~60fps which is approximatively 16.66ms, not interesting to throttle under that value
+    // because we would probably resend the same frame again.
```
Comment (vkrause) [inline core/remoteviewserver.cpp]: That is actually wrong, it does not resend in that case. It also checks if the source actually changed. At least that's what it is supposed to be doing, would be nice to test as well I guess. Ie. if you have a static scene, it should only send you one frame, until you tell it the scene has changed.

File: common/remoteviewinterface.h [hunk 5]
```diff
@@ -76,6 +76,8 @@ public slots:
 
     virtual void setViewActive(bool active) = 0;
 
+    virtual void setUpdateInterval(int ms) = 0;
```
Comment (vkrause) [inline common/remoteviewinterface.h]: This is not something anyone externally should control, it's purely an implementation detail to reduce event loop load. Can we get away without having this somehow?

File: plugins/quickinspector/quickoverlay.cpp [hunk 6]
```diff
@@ -437,16 +439,16 @@ void QuickOverlay::windowAfterRendering()
         m_grabbedFrame.transform.translate(intersect.x() * m_renderInfo.dpr , -intersect.y() * m_renderInfo.dpr - h);
         m_grabbedFrame.image.setDevicePixelRatio(m_renderInfo.dpr);
 
-        if (!m_grabbedFrame.image.isNull()) {
-            m_sceneGrabbed.invoke(this, Qt::QueuedConnection, Q_ARG(GammaRay::GrabbedFrame, m_grabbedFrame));
-        }
+        Q_ASSERT(!m_grabbedFrame.image.isNull());
```
Comment (vkrause) [inline plugins/quickinspector/quickoverlay.cpp]: Sure this is always the case? Can w and h be 0 above?

File: plugins/quickinspector/quickoverlay.cpp [hunk 7]
```diff
@@ -437,16 +439,16 @@ void QuickOverlay::windowAfterRendering()
...
+        m_sceneGrabbed.invoke(this, Qt::QueuedConnection, Q_ARG(GammaRay::GrabbedFrame, m_grabbedFrame));
     }
 
     drawDecorations();
 
     m_window->resetOpenGLState();
 
     if (m_isGrabbingMode) {
+        locker.unlock();
         setIsGrabbingMode(false);
```
Comment (vkrause) [inline plugins/quickinspector/quickoverlay.cpp]: Seeing this pattern also below, wouldn't it be better to assume the lock being held in setIsGrabbingMode? So the value doesn't change between the check here, and the call to it?

File: tests/actiontest.cpp (hunk 1 above)
Comment (pasnox) [inline tests/actiontest.cpp]: I think no because there is a macro inside it meaning it can be built 2 differents way.
Or I would have to dissociate both *probe* ways as different classes, then probably I can do that.

Comment (vkrause) [inline tests/actiontest.cpp]: Is that distinction actually necessary? or is that just because one test was written somewhat differently for historic reasons?

Comment (pasnox) [inline tests/actiontest.cpp]: Looks like we can merge the code path and go that way. Will do.

File: core/remoteviewserver.cpp (hunk 2 above)
Comment (pasnox) [inline core/remoteviewserver.cpp]: Yep, though the signal is not part of the interface it's why I added a new one.
In the end I can indeed use this one using Qt 4 style connection.
Fixed

File: core/remoteviewserver.cpp (hunk 4 above)
Comment (pasnox) [inline core/remoteviewserver.cpp]: It was in the expectation that we would force request it, in this case unless for Qt 3D scene, we would not have a changed result frame, will change the comment.
I will add a test for that as well.

File: common/remoteviewinterface.h (hunk 5 above)
Comment (pasnox) [inline common/remoteviewinterface.h]: Fixed

File: plugins/quickinspector/quickoverlay.cpp (hunk 6 above)
Comment (pasnox) [inline plugins/quickinspector/quickoverlay.cpp]: Fixed

File: plugins/quickinspector/quickoverlay.cpp (hunk 7 above)
Comment (pasnox) [inline plugins/quickinspector/quickoverlay.cpp]: Fixed

File: tests/actiontest.cpp (hunk 1 above)
Comment (pasnox) [inline tests/actiontest.cpp]: Fixed

File: core/remoteviewserver.cpp (hunk 4 above)
Comment (pasnox) [inline core/remoteviewserver.cpp]: Fixed

File: tests/quickinspectortest.cpp (hunk 3 above)
Comment (pasnox) [inline tests/quickinspectortest.cpp]: Fixed

File: plugins/quickinspector/quickoverlay.cpp [hunk 8]
```diff
@@ -437,17 +441,18 @@ void QuickOverlay::windowAfterRendering()
         m_grabbedFrame.transform.translate(intersect.x() * m_renderInfo.dpr , -intersect.y() * m_renderInfo.dpr - h);
         m_grabbedFrame.image.setDevicePixelRatio(m_renderInfo.dpr);
 
-        if (!m_grabbedFrame.image.isNull()) {
-            m_sceneGrabbed.invoke(this, Qt::QueuedConnection, Q_ARG(GammaRay::GrabbedFrame, m_grabbedFrame));
-        }
+        // Let emit the signal even if our image is possibly null, this way we make perfect ping/pong
+        // reuests making it easier to unit test.
```
Comment (vkrause) [inline plugins/quickinspector/quickoverlay.cpp]: Easier to unit test is not the design criteria here ;-) Can the client handle empty frames? How often does  this happen, ie. is this causing unreasonable performance costs?

File: tests/CMakeLists.txt [hunk 9]
```diff
@@ -13,20 +14,76 @@ else()
...
 endif()
 
+# create intermediate base_test_obj object files which unit tests can use
+add_library(base_test_obj OBJECT basetest.h)
+set_target_properties(base_test_obj PROPERTIES POSITION_INDEPENDENT_CODE ON)
+target_compile_definitions(base_test_obj PRIVATE ${Qt5Core_DEFINITIONS} ${Qt5Test_DEFINITIONS}) # for -DQT_NAMESPACE=...
+target_include_directories(base_test_obj SYSTEM PRIVATE
+    ${Qt5Core_INCLUDE_DIRS}
+    ${Qt5Test_INCLUDE_DIRS}
+)
```
Comment (vkrause) [inline tests/CMakeLists.txt]: This looks like it might not build with Qt4?

File: tests/CMakeLists.txt [hunk 10]
```diff
@@ -13,20 +14,76 @@ else()
...
+set_target_properties(base_probe_test_obj PROPERTIES POSITION_INDEPENDENT_CODE ON)
+target_compile_definitions(base_probe_test_obj PRIVATE ${Qt5Core_DEFINITIONS} ${Qt5Test_DEFINITIONS}) # for -DQT_NAMESPACE=...
+target_include_directories(base_probe_test_obj PUBLIC
+    $<TARGET_PROPERTY:gammaray_core,INTERFACE_INCLUDE_DIRECTORIES>
+    $<TARGET_PROPERTY:gammaray_common,INTERFACE_INCLUDE_DIRECTORIES>
+)
+target_include_directories(base_probe_test_obj SYSTEM PRIVATE
+    ${Qt5Core_INCLUDE_DIRS}
+    ${Qt5Test_INCLUDE_DIRS}
+)
```
Comment (vkrause) [inline tests/CMakeLists.txt]: same here

File: tests/basetest.h [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+
+#include <config-gammaray.h>
+
+#include <QObject>
+#include <QSignalSpy>
+#include <QAbstractItemModel>
+
+#include <QtTest/qtest.h>
+
+class BaseTest : public QObject
```
Comment (vkrause) [inline tests/basetest.h]: Good idea to factor out those utilities. They don't need to be in the inheritance hierarchy, right? So we could have them in a separate utility namespace for example, would make it easier to combine this with the other base classes.

File: tests/quickinspectortest.cpp [hunk 12]
```diff
@@ -258,18 +203,123 @@ private slots:
...
+            rootItem->setProperty("interval", qRound(animationInterval));
+            rootItem->setProperty("animated", true);
+            QTest::qWait(qRound(animationDuration));
+
+            // Wait to process pending requests...
+            rootItem->setProperty("animated", false);
+            QTest::qWait(qRound(animationDuration));
+
+            QVERIFY(requestedSpy.count() <= qRound(maxPossibleThrottledRequests * 1.05));
+            QVERIFY(updatedSpy.count() <= qRound(maxPossibleThrottledRequests * 1.05));
```
Comment (vkrause) [inline tests/quickinspectortest.cpp]: We never call clientViewUpdated() here, so shouldn't this be always 1 here, no matter how much the source actually changed? That would actually be a valid test-case, independent of what I suspect this code tried to test :)

File: common/remoteviewinterface.h [hunk 13]
```diff
@@ -85,6 +85,8 @@ public slots:
     void reset();
     void elementsAtReceived(const GammaRay::ObjectIds &ids, int bestCandidate);
     void frameUpdated(const GammaRay::RemoteViewFrame &frame);
+    /// when server receive this signal, obtain a new frame and send it to the client
+    void requestUpdate();
```
Comment (pasnox) [inline common/remoteviewinterface.h]: Hm, I can move it back to server again , that's indeed a performance penalty.
The issue was that I then can't do any Qt5 connect style to a lambda.
With Qt 4 connect style I would need a slot, being *checked* by the unit test framework.

File: core/remoteviewserver.cpp [hunk 14]
```diff
@@ -57,7 +57,9 @@ RemoteViewServer::RemoteViewServer(const QString &name, QObject *parent)
                                                     name), this, "clientConnectedChanged");
 
     m_updateTimer->setSingleShot(true);
-    m_updateTimer->setInterval(10);
+    // Qml try to render @ ~60fps which is approximatively 16.66ms, not interesting to throttle
+    // under that value.
+    m_updateTimer->setInterval(qRound(1000.0 / 60.0));
```
Comment (pasnox) [inline core/remoteviewserver.cpp]: Fixed

File: core/remoteviewserver.cpp [hunk 15]
```diff
@@ -91,7 +93,6 @@ bool RemoteViewServer::isActive() const
 
 void RemoteViewServer::sendFrame(const RemoteViewFrame &frame)
 {
-    m_clientReady = false;
```
Comment (pasnox) [inline core/remoteviewserver.cpp]: The reason was specified in the commit message as "It was possible to trigger a new update request if sourceChanged
was called in between requestUpdateTimeout and sendFrame."
The sourceChanged being directly connected to the QuickOverlay::sceneChanged() signal.
In rare case the signal is emitted in between the 2 quoted calls which trigger a non wanted update request.

File: plugins/quickinspector/quickoverlay.cpp (hunk 8 above)
Comment (pasnox) [inline plugins/quickinspector/quickoverlay.cpp]: Yes it can handle null images.
Technically I never had a case of null image, but I suppose it can arrive if the window root item has a size of 0x0 (yeah that's improbable case but still possible).
Does cost nothing as far as I can see.

File: core/remoteviewserver.cpp (hunk 15 above)
Comment (vkrause) [inline core/remoteviewserver.cpp]: But this was correct, no? If the source changed after we grabbed the frame we eventually need to get a second frame. After all we want to show the correct state in the client.

Comment (pasnox) [inline core/remoteviewserver.cpp]: Hm I don't think so.
If we reach requestUpdateTimeout a QuickOverlay grab request is triggered.
Depending at which time the request is done, we can already be in locked state in QuickOverlay::windowAfterRendering which will emit a sceneChanged just before we can change the grabbing mode. This will then trigger a new update request while we did not even yet changed the QuickOverlay grabbing mode. As a result the next request timeout will try to request a frame grab but **could** do nothing because we **could** be in the process of capturing a frame already (ie, grabbing mode is true already).
Schematized by: "Request -> Request -> Grab" **instead of** "Request -> Grab -> Request -> Grab"
So for me if the client requested a frame but did not yet received it, I think a request update can't occurs server side as normally only a clientViewUpdated call could trigger a possible frame grab (assaying - client is ready).
As a consequence, requesting a frame update should mark client as non ready.
If my reasoning sound wrong to you, then probably we want to fix the double update request calls (ie, by introducing a quickoverlay grabbing state test in RemoteViewServer::checkRequestUpdate() so the new request will only occurs after the current grab finished).

Comment (pasnox) [inline core/remoteviewserver.cpp]: After writing all that it make it clear that the problem comes from the differences between the different grabbers... widget grabber is synchrone while the quick one is asynchrone now (used to be synchrone before but no longer). So we did not get that corner case before.

File: tests/quickinspectortest.cpp (hunk 12 above)
Comment (pasnox) [inline tests/quickinspectortest.cpp]: We do call it, from the upper connection.

File: plugins/quickinspector/quickoverlay.cpp (hunk 8 above)
Comment (pasnox) [inline plugins/quickinspector/quickoverlay.cpp]: Fixed

File: tests/CMakeLists.txt (hunk 9 above)
Comment (pasnox) [inline tests/CMakeLists.txt]: Fixed

File: tests/CMakeLists.txt (hunk 10 above)
Comment (pasnox) [inline tests/CMakeLists.txt]: Fixed

File: tests/basetest.h (hunk 11 above)
Comment (pasnox) [inline tests/basetest.h]: Fixed

File: tests/quickinspectortest.cpp (hunk 12 above)
Comment (pasnox) [inline tests/quickinspectortest.cpp]: Fixed

File: core/remoteviewserver.cpp (hunk 15 above)
Comment (pasnox) [inline core/remoteviewserver.cpp]: Fixed
